File: Rosella/src/main/java/me/hydos/rosella/Rosella.java
package me.hydos.rosella;

import me.hydos.rosella.device.LegacyVulkanDevice;
import me.hydos.rosella.device.VulkanDevice;
import me.hydos.rosella.device.VulkanQueues;
import me.hydos.rosella.display.Display;
import me.hydos.rosella.device.init.DeviceBuilder;
import me.hydos.rosella.device.init.InitializationRegistry;
import me.hydos.rosella.device.init.InstanceBuilder;
import me.hydos.rosella.device.init.VulkanInstance;
import me.hydos.rosella.device.init.features.RosellaLegacy;
import me.hydos.rosella.logging.DebugLogger;
import me.hydos.rosella.logging.DefaultDebugLogger;
import me.hydos.rosella.memory.ThreadPoolMemory;
import me.hydos.rosella.memory.buffer.GlobalBufferManager;
import me.hydos.rosella.render.renderer.Renderer;
import me.hydos.rosella.scene.object.ObjectManager;
import me.hydos.rosella.scene.object.impl.SimpleObjectManager;
import me.hydos.rosella.util.SemaphorePool;
import me.hydos.rosella.vkobjects.VkCommon;
import me.hydos.rosella.vkobjects.LegacyVulkanInstance;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.message.StringFormatterMessageFactory;
import org.lwjgl.system.MemoryStack;
import org.lwjgl.vulkan.VK10;
import org.lwjgl.vulkan.VkLayerProperties;

import java.nio.IntBuffer;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static me.hydos.rosella.util.VkUtils.ok;
import static org.lwjgl.vulkan.EXTDebugUtils.vkDestroyDebugUtilsMessengerEXT;
import static org.lwjgl.vulkan.KHRSurface.vkDestroySurfaceKHR;
import static org.lwjgl.vulkan.VK10.*;
import static org.lwjgl.vulkan.VK12.VK_API_VERSION_1_2;

/**
 * Main Rosella class. If you're interacting with the engine from here, You will most likely be safe.
 */
public class Rosella {

    public static final Logger LOGGER = LogManager.getLogger("Rosella", new StringFormatterMessageFactory());
    public static final int VULKAN_VERSION = VK_API_VERSION_1_2;
    public final GlobalBufferManager bufferManager;
    public final VkCommon common = new VkCommon();
    public final Renderer renderer;
    public final ObjectManager objectManager;

    public final VulkanInstance vulkanInstance;
    public final VulkanDevice vulkanDevice;

    public Rosella(InitializationRegistry registry, Display display, String applicationName, int applicationVersion) {
        // TODO remove
        registry.enableValidation(true);

        common.display = display;
        display.getRequiredExtensions().forEach(registry::addRequiredInstanceExtensions);

        // Needed because debug callbacks are handled by LegacyVulkanInstance. TODO remove this
        common.vkInstance = new LegacyVulkanInstance(registry, applicationName, applicationVersion, new DefaultDebugLogger());
        this.vulkanInstance = common.vkInstance.newInstance;

        common.surface = display.createSurface(common);
        registry.registerApplicationFeature(new RosellaLegacy(common));
        registry.addRequiredApplicationFeature(RosellaLegacy.NAME);

        this.vulkanDevice = new DeviceBuilder(this.vulkanInstance, registry).build();
        common.device = new LegacyVulkanDevice(this.vulkanDevice);

        RosellaLegacy.RosellaLegacyFeatures legacyFeatures = RosellaLegacy.getMetadata(this.vulkanDevice);
        try {
            common.queues = new VulkanQueues(legacyFeatures.graphicsQueue().get(), legacyFeatures.presentQueue().get());
        } catch (Exception ex) {
            throw new RuntimeException("Not good stuff.");
        }

        // TODO: Tons and tons of old code. Need to remove
        common.memory = new ThreadPoolMemory(common);
        common.semaphorePool = new SemaphorePool(common.device.rawDevice);

        this.objectManager = new SimpleObjectManager(this, common);
        this.renderer = new Renderer(this);
        ((SimpleObjectManager) objectManager).textureManager.initializeBlankTexture(renderer);
        this.objectManager.postInit(renderer);
        this.bufferManager = new GlobalBufferManager(this);

        display.onReady();
    }

    @Deprecated
    public Rosella(Display display, String applicationName, boolean enableBasicValidation) {
        this(display, enableBasicValidation ? Collections.singletonList("VK_LAYER_KHRONOS_validation") : Collections.emptyList(), applicationName, new DefaultDebugLogger());
    }

    @Deprecated
    public Rosella(Display display, List<String> requestedValidationLayers, String applicationName, DebugLogger debugLogger) {
        List<String> requiredExtensions = display.getRequiredExtensions();

        InitializationRegistry initializationRegistry = new InitializationRegistry();
        requestedValidationLayers.forEach(initializationRegistry::addRequiredInstanceLayer);
        requiredExtensions.forEach(initializationRegistry::addRequiredInstanceExtensions);

        // Setup core vulkan stuff
        common.display = display;
        common.vkInstance = new LegacyVulkanInstance(initializationRegistry, applicationName, VK10.VK_MAKE_VERSION(1, 0, 0), debugLogger);

        common.surface = display.createSurface(common);
        initializationRegistry.registerApplicationFeature(new RosellaLegacy(common));

        common.device = new LegacyVulkanDevice(common.vkInstance.newInstance, initializationRegistry);

        RosellaLegacy.RosellaLegacyFeatures legacyFeatures = RosellaLegacy.getMetadata(common.device.newDevice);
        try {
            common.queues = new VulkanQueues(legacyFeatures.graphicsQueue().get(), legacyFeatures.presentQueue().get());
        } catch (Exception ex) {
            throw new RuntimeException("Not good stuff.");
        }

        common.memory = new ThreadPoolMemory(common);
        common.semaphorePool = new SemaphorePool(common.device.rawDevice);

        // Setup the object manager
        this.objectManager = new SimpleObjectManager(this, common);
        this.renderer = new Renderer(this); //TODO: make swapchain, etc initialization happen outside of the renderer and in here
        ((SimpleObjectManager) objectManager).textureManager.initializeBlankTexture(renderer); // TODO: move this maybe
        this.objectManager.postInit(renderer);
        this.bufferManager = new GlobalBufferManager(this);

        // Tell the display we are initialized
        display.onReady();

        this.vulkanInstance = common.vkInstance.newInstance;
        this.vulkanDevice = common.device.newDevice;
    }

    /**
     * Free's the vulkan resources.
     */
    public void free() {
        common.device.waitForIdle();
        objectManager.free();
        renderer.free();

        // Free the rest of it
        common.memory.free();
        common.semaphorePool.free();

        vkDestroyCommandPool(common.device.rawDevice, renderer.commandPool, null);

        vulkanDevice.destroy();

        vkDestroySurfaceKHR(common.vkInstance.rawInstance, common.surface, null);

        common.vkInstance.messenger.ifPresent(messenger -> { // FIXME
            vkDestroyDebugUtilsMessengerEXT(common.vkInstance.rawInstance, messenger, null);
        });

        vulkanInstance.destroy();

        common.display.exit();
    }

    /**
     * Checks if the system supports validation layers.
     *
     * @param requestedValidationLayers the validation layers requested by the application/user
     * @return if the system supports the request validation layers.
     */
    private boolean validationLayersSupported(List<String> requestedValidationLayers) {
        return getSupportedValidationLayers().containsAll(requestedValidationLayers);
    }

    /**
     * Gets all validation layers supported by the machine
     *
     * @return all validation layers that are supported
     */
    private Set<String> getSupportedValidationLayers() {
        try (MemoryStack stack = MemoryStack.stackPush()) {
            IntBuffer pLayerCount = stack.ints(0);
            ok(vkEnumerateInstanceLayerProperties(pLayerCount, null));
            VkLayerProperties.Buffer availableLayers = VkLayerProperties.mallocStack(pLayerCount.get(0), stack);
            ok(vkEnumerateInstanceLayerProperties(pLayerCount, availableLayers));
            return availableLayers.stream()
                    .map(VkLayerProperties::layerNameString)
                    .collect(Collectors.toSet());
        }

    }

    static {
        LOGGER.atLevel(Level.ALL);
    }
}

File: Rosella/src/main/kotlin/me/hydos/rosella/Rosella.kt
package me.hydos.rosella

import me.hydos.rosella.audio.SoundManager
import me.hydos.rosella.render.camera.Camera
import me.hydos.rosella.render.device.Device
import me.hydos.rosella.render.io.Window
import me.hydos.rosella.render.material.Material
import me.hydos.rosella.render.model.RenderObject
import me.hydos.rosella.render.renderer.Renderer
import me.hydos.rosella.render.resource.Identifier
import me.hydos.rosella.render.shader.RawShaderProgram
import me.hydos.rosella.render.shader.ShaderManager
import me.hydos.rosella.render.swapchain.Frame
import me.hydos.rosella.render.texture.TextureManager
import me.hydos.rosella.render.util.memory.Memory
import me.hydos.rosella.render.util.ok
import org.lwjgl.PointerBuffer
import org.lwjgl.glfw.GLFW.glfwShowWindow
import org.lwjgl.glfw.GLFWVulkan
import org.lwjgl.glfw.GLFWVulkan.glfwCreateWindowSurface
import org.lwjgl.system.MemoryStack.stackGet
import org.lwjgl.system.MemoryStack.stackPush
import org.lwjgl.system.MemoryUtil.NULL
import org.lwjgl.vulkan.*
import org.lwjgl.vulkan.KHRSurface.vkDestroySurfaceKHR
import org.lwjgl.vulkan.VK10.*
import org.lwjgl.vulkan.VK12.VK_API_VERSION_1_2
import java.nio.IntBuffer
import java.nio.LongBuffer
import java.util.function.Consumer
import java.util.stream.Collectors

/**
 * Main engine class. most interactions will happen here
 */
class Rosella(
	name: String,
	val enableValidationLayers: Boolean,
	val window: Window
) {
	val polygonMode: Int = VK_POLYGON_MODE_FILL
	var memory: Memory

	var renderer: Renderer = Renderer()

	var renderObjects = HashMap<String, RenderObject>()
	var materials = HashMap<Identifier, Material>()
	var shaderManager: ShaderManager
	var textureManager: TextureManager

	val camera = Camera(window)

	internal lateinit var vulkanInstance: VkInstance
	lateinit var maxImages: IntBuffer

	val device: Device
	private var debugMessenger: Long = 0
	var surface: Long = 0

	init {
		SoundManager.initialize()
		window.onWindowResize(renderer::windowResizeCallback)

		val validationLayers = defaultValidationLayers.toSet()
		if (enableValidationLayers && !validationLayersSupported(validationLayers)) {
			throw RuntimeException("Validation Layers are not available!")
		}

		createInstance(name, validationLayers)

		if (enableValidationLayers) {
			setupDebugMessenger()
		}

		createSurface()
		this.device = Device(this, validationLayers)
		this.shaderManager = ShaderManager(device)
		this.textureManager = TextureManager(device)
		this.memory = Memory(device, vulkanInstance)
		renderer.initialize(this)

		glfwShowWindow(window.windowPtr)
	}

	private fun createInstance(name: String, validationLayers: Set<String>) {
		stackPush().use { stack ->
			val applicationInfo = VkApplicationInfo.callocStack(stack)
				.sType(VK_STRUCTURE_TYPE_APPLICATION_INFO)
				.pApplicationName(stack.UTF8Safe(name))
				.applicationVersion(VK_MAKE_VERSION(1, 0, 0))
				.pEngineName(stack.UTF8Safe("Rosella"))
				.engineVersion(VK_MAKE_VERSION(0, 1, 0))
				.apiVersion(VK_API_VERSION_1_2)
			val createInfo = VkInstanceCreateInfo.callocStack(stack)
				.pApplicationInfo(applicationInfo)
				.sType(VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO)
				.ppEnabledExtensionNames(getRequiredExtensions(enableValidationLayers))
			if (enableValidationLayers) {
				createInfo.ppEnabledLayerNames(asPtrBuffer(validationLayers))
				val debugCreateInfo = VkDebugUtilsMessengerCreateInfoEXT.callocStack(stack)
				populateDebugMessengerCreateInfo(debugCreateInfo)
				createInfo.pNext(debugCreateInfo.address())
			}

			val pInstance = stack.mallocPointer(1)
			vkCreateInstance(createInfo, null, pInstance).ok()

			vulkanInstance = VkInstance(pInstance[0], createInfo)
		}
	}

	private fun createSurface() {
		stackPush().use {
			val pSurface: LongBuffer = it.longs(VK_NULL_HANDLE)
			glfwCreateWindowSurface(vulkanInstance, window.windowPtr, null, pSurface).ok()
			this.surface = pSurface.get(0)
		}
	}

	fun free() {
		for (model in renderObjects.values) {
			model.free(memory)
		}

		renderer.freeSwapChain(this)

		renderer.inFlightFrames.forEach(Consumer { frame: Frame ->
			vkDestroySemaphore(device.device, frame.renderFinishedSemaphore(), null)
			vkDestroySemaphore(device.device, frame.imageAvailableSemaphore(), null)
			vkDestroyFence(device.device, frame.fence(), null)
		})

		vkDestroyCommandPool(device.device, renderer.commandPool, null)

		renderer.swapChain.free(device.device)

		vkDestroyDevice(device.device, null)

		if (vkGetInstanceProcAddr(vulkanInstance, "vkDestroyDebugUtilsMessengerEXT") != NULL) {
			EXTDebugUtils.vkDestroyDebugUtilsMessengerEXT(vulkanInstance, debugMessenger, null)
		}

		vkDestroySurfaceKHR(vulkanInstance, surface, null)
		vkDestroyInstance(vulkanInstance, null)

		memory.free()
	}

	private fun getRequiredExtensions(validationLayersEnabled: Boolean): PointerBuffer? {
		val glfwExtensions = GLFWVulkan.glfwGetRequiredInstanceExtensions()
		if (validationLayersEnabled) {
			val stack = stackGet()
			val extensions = stack.mallocPointer(glfwExtensions!!.capacity() + 1)
			extensions.put(glfwExtensions)
			extensions.put(stack.UTF8(EXTDebugUtils.VK_EXT_DEBUG_UTILS_EXTENSION_NAME))
			return extensions.rewind()
		}
		return glfwExtensions
	}

	private val defaultValidationLayers: List<String>
		get() {
			val validationLayers: MutableList<String> = ArrayList()
			validationLayers.add("VK_LAYER_KHRONOS_validation")
			return validationLayers
		}

	private fun setupDebugMessenger() {
		stackPush().use { stack ->
			val createInfo = VkDebugUtilsMessengerCreateInfoEXT.callocStack(stack)
			populateDebugMessengerCreateInfo(createInfo)
			val pDebugMessenger = stack.longs(VK_NULL_HANDLE)
			if (createDebugUtilsMessengerEXT(vulkanInstance, createInfo, null, pDebugMessenger) != VK_SUCCESS) {
				throw RuntimeException("Failed to set up debug messenger")
			}
			debugMessenger = pDebugMessenger[0]
		}
	}

	private fun populateDebugMessengerCreateInfo(debugCreateInfo: VkDebugUtilsMessengerCreateInfoEXT) {
		debugCreateInfo.sType(EXTDebugUtils.VK_STRUCTURE_TYPE_DEBUG_UTILS_MESSENGER_CREATE_INFO_EXT)
			.messageSeverity(EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT or EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT or EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT)
			.messageType(EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_TYPE_GENERAL_BIT_EXT or EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_TYPE_VALIDATION_BIT_EXT or EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT)
			.pfnUserCallback(this::debugCallback)
	}

	private fun debugCallback(severity: Int, messageType: Int, pCallbackData: Long, pUserData: Long): Int {
		val callbackData = VkDebugUtilsMessengerCallbackDataEXT.create(pCallbackData)
		val message = callbackData.pMessageString()
		if (severity == EXTDebugUtils.VK_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT) {
			if(message.startsWith("Validation Error")) {
				val split = message.split("|")
				val cause = split[2]
				RuntimeException(cause).printStackTrace()
			} else{
				System.err.println(message)
			}
		} else {
			println(message)
		}
		return VK_FALSE
	}

	internal fun asPtrBuffer(validationLayers: Set<String>): PointerBuffer {
		val stack = stackGet()
		val buffer = stack.mallocPointer(validationLayers.size)
		for (validationLayer in validationLayers) {
			val byteBuffer = stack.UTF8(validationLayer)
			buffer.put(byteBuffer)
		}
		return buffer.rewind()
	}

	private fun validationLayersSupported(validationLayers: Set<String>): Boolean {
		stackPush().use { stack ->
			val layerCount = stack.ints(0)
			vkEnumerateInstanceLayerProperties(layerCount, null).ok()
			val availableLayers = VkLayerProperties.mallocStack(layerCount[0], stack)
			vkEnumerateInstanceLayerProperties(layerCount, availableLayers).ok()
			val availableLayerNames = availableLayers.stream()
				.map { obj: VkLayerProperties -> obj.layerNameString() }
				.collect(Collectors.toSet())
			return availableLayerNames.containsAll(validationLayers)
		}
	}

	fun addRenderObject(renderObject: RenderObject, name: String) {
		if(renderObjects.containsKey(name)) {
			error("An render object already exists with that name!")
		}
		renderObject.load(this)
		renderObjects[name] = renderObject
		renderObject.create(this)
	}

	fun registerMaterial(identifier: Identifier, material: Material) {
		materials[identifier] = material
	}

	fun registerShader(identifier: Identifier, rawShader: RawShaderProgram) {
		shaderManager.shaders[identifier] = rawShader
	}

	fun reloadMaterials() {
		var test = 0
		for (material in materials.values) {
			material.loadShaders(this)
			material.loadTextures(this)
			material.shader.raw.createDescriptorSetLayout()
			material.createPipeline(
				device,
				renderer.swapChain,
				renderer.renderPass,
				material.shader.raw.descriptorSetLayout,
				polygonMode
			)
			test++
			println(test.toString() + " / " + materials.values.size)
		}
		println("Recreating Swap Chain")
		renderer.recreateSwapChain(window, camera, this)
		println("Swapchain Recreated")
	}

	private fun createDebugUtilsMessengerEXT(
		instance: VkInstance,
		createInfo: VkDebugUtilsMessengerCreateInfoEXT,
		allocationCallbacks: VkAllocationCallbacks?,
		pDebugMessenger: LongBuffer
	): Int {
		return if (vkGetInstanceProcAddr(instance, "vkCreateDebugUtilsMessengerEXT") != NULL) {
			EXTDebugUtils.vkCreateDebugUtilsMessengerEXT(instance, createInfo, allocationCallbacks, pDebugMessenger)
		} else VK_ERROR_EXTENSION_NOT_PRESENT
	}

	fun getHeight(): Float {
		return renderer.swapChain.swapChainExtent.height().toFloat()
	}

	fun getWidth(): Float {
		return renderer.swapChain.swapChainExtent.width().toFloat()
	}
}
